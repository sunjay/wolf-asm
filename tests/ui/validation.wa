# Statement before any sections
.b1 33
add $1, $2

# Label before any sections is fine
beforeeverythingwooooo:

# Wrong order of sections
section .static

.const a 32

# Duplicate label name
foo: foo:
  .b8 32323232

  # Value out of range
  .b1 256
  .b2 65536
  .b4 4294967296

  # Min/Max values should be fine
  .b1 0
  .b2 0
  .b4 0
  .b8 0
  .b1 0xff
  .b2 0xffff
  .b4 0xffff_ffff
  .b8 0xffff_ffff_ffff_ffff

# Duplicate constant with the same value is allowed and NOT an error
.const a 32

# Duplicate constant with a different value (warning)
.const a 33

bar:
  # This will get the value 33 for `a`
  add $1, a

  # Conflict with label name
  .const bar 981

# Duplicate label name
foo:

  # Negative size
  .zero -12
  .uninit -1

section .code

  # Unknown instruction
  foo $1

  # Valid register
  pop $0
  pop $32
  pop $63
  pop $sp
  pop $fp
  # Invalid register
  pop $64
  pop $128
  pop $255
  pop $foo
  pop $bp

  add $1, a # valid
  add a, $1 # immediate not allowed as destination
  add 123, $1 # immediate not allowed as destination
  add # no arguments
  add $1 # one argument
  add $1, $2, $3 # extra argument
  add 123, $1, $2, $3 # immediate as destination and extra arguments

  load1 $1, 0x1 # valid
  load1 $1, a # valid
  load1 $1, $2 # valid
  load1 0x1, $2 # immediate not allowed as destination
  load1 a, $2 # immediate not allowed as destination
  load1 # no arguments
  load1 $1 # one argument
  load1 $21, $22, $23 # extra argument
  load1 123, $1, $2, $3 # immediate as destination and extra arguments

  load1 $1, -4($2) # valid
  load1 $1, 48($2) # valid
  load1 $1, -32768($2) # valid (min offset)
  load1 $1, 32767($2) # valid (max offset)
  load1 $1, -32769($2) # offset too large (negative)
  load1 $1, 32768($2) # offset too large (positive)

  mul $2, $1 # valid
  mull $3, $2, $1 # valid (extra arg)
  mulu $2, $1 # valid
  mullu $3, $2, $1 # valid (extra arg)
  mul # no arguments
  mull # no arguments
  mulu # no arguments
  mullu # no arguments
  mul $3, $2, $1 # extra argument
  mull $3, $2, $1, $4 # extra argument
  mulu $3, $2, $1 # extra argument
  mullu $3, $2, $1, $4 # extra argument

  nop # valid
  nop $1 # extra argument
  nop $1, $2 # extra arguments

# Duplicate section
section .static

  # Empty strings are allowed and NOT an error
  .bytes ''
  .bytes ""

# Duplicate section
section .code
