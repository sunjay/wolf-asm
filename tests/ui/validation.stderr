[../tests/ui/validation.wa:14:6-14:8] error: duplicate label name `foo`
[../tests/ui/validation.wa:14:1-14:3] note: originally defined here

[../tests/ui/validation.wa:46:1-46:3] error: duplicate label name `foo`
[../tests/ui/validation.wa:14:1-14:3] note: originally defined here

[../tests/ui/validation.wa:33:1-33:11] warning: constant named `a` was redefined
[../tests/ui/validation.wa:11:1-11:11] note: the previous declaration of this constant

[../tests/ui/validation.wa:36:1-36:11] warning: constant named `a` was redefined
[../tests/ui/validation.wa:33:1-33:11] note: the previous declaration of this constant

[../tests/ui/validation.wa:43:3-43:16] error: constant name `bar` conflicts with a label name
[../tests/ui/validation.wa:38:1-38:3] note: the conflicting label

[../tests/ui/validation.wa:2:1-2:6] error: all assembly statements must occur within a section, e.g. `section .code`

[../tests/ui/validation.wa:3:1-3:10] error: all assembly statements must occur within a section, e.g. `section .code`

[../tests/ui/validation.wa:18:3-18:9] error: value `256` for `.b1` must be in the range `0` to `255`

[../tests/ui/validation.wa:19:3-19:11] error: value `65536` for `.b2` must be in the range `0` to `65535`

[../tests/ui/validation.wa:20:3-20:16] error: value `4294967296` for `.b4` must be in the range `0` to `4294967295`

[../tests/ui/validation.wa:49:9-49:11] error: number of bytes must be non-negative

[../tests/ui/validation.wa:50:11-50:12] error: number of bytes must be non-negative

[../tests/ui/validation.wa:52:1-52:13] error: the `.code` section must occur before the `.static` section

[../tests/ui/validation.wa:55:3-55:5] error: unknown instruction `foo`

[../tests/ui/validation.wa:64:7-64:9] error: invalid register `$64`
[../tests/ui/validation.wa:64:7-64:9] note: registers must be `$0` to `$63`, `$sp`, or `$fp`

[../tests/ui/validation.wa:65:7-65:10] error: invalid register `$128`
[../tests/ui/validation.wa:65:7-65:10] note: registers must be `$0` to `$63`, `$sp`, or `$fp`

[../tests/ui/validation.wa:66:7-66:10] error: invalid register `$255`
[../tests/ui/validation.wa:66:7-66:10] note: registers must be `$0` to `$63`, `$sp`, or `$fp`

[../tests/ui/validation.wa:67:7-67:10] error: invalid register `$foo`
[../tests/ui/validation.wa:67:7-67:10] note: registers must be `$0` to `$63`, `$sp`, or `$fp`

[../tests/ui/validation.wa:68:7-68:9] error: invalid register `$bp`
[../tests/ui/validation.wa:68:7-68:9] note: registers must be `$0` to `$63`, `$sp`, or `$fp`

[../tests/ui/validation.wa:71:7] error: expected a register, found `33`

[../tests/ui/validation.wa:72:7-72:9] error: expected a register, found `123`

[../tests/ui/validation.wa:73:3-73:5] error: expected a destination argument for `add` instruction (takes 2 arguments)

[../tests/ui/validation.wa:73:3-73:5] error: expected a source argument for `add` instruction (takes 2 arguments)

[../tests/ui/validation.wa:74:3-74:5] error: expected a source argument for `add` instruction (takes 2 arguments)

[../tests/ui/validation.wa:75:3-75:5] error: expected 2 arguments for `add` instruction, found 3 arguments

[../tests/ui/validation.wa:76:7-76:9] error: expected a register, found `123`

[../tests/ui/validation.wa:76:3-76:5] error: expected 2 arguments for `add` instruction, found 4 arguments

[../tests/ui/validation.wa:81:9-81:11] error: expected a register, found `1`

[../tests/ui/validation.wa:82:9] error: expected a register, found `33`

[../tests/ui/validation.wa:83:3-83:7] error: expected a destination argument for `load1` instruction (takes 2 arguments)

[../tests/ui/validation.wa:83:3-83:7] error: expected a location argument for `load1` instruction (takes 2 arguments)

[../tests/ui/validation.wa:84:3-84:7] error: expected a location argument for `load1` instruction (takes 2 arguments)

[../tests/ui/validation.wa:85:3-85:7] error: expected 2 arguments for `load1` instruction, found 3 arguments

[../tests/ui/validation.wa:86:9-86:11] error: expected a register, found `123`

[../tests/ui/validation.wa:86:3-86:7] error: expected 2 arguments for `load1` instruction, found 4 arguments

[../tests/ui/validation.wa:92:13-92:18] error: offset value `-32769` must be in the range of a 16-bit signed integer, `-32768` to `32767`

[../tests/ui/validation.wa:93:13-93:17] error: offset value `32768` must be in the range of a 16-bit signed integer, `-32768` to `32767`

[../tests/ui/validation.wa:99:3-99:5] error: expected a destination argument for `mul` instruction (takes 2 arguments)

[../tests/ui/validation.wa:99:3-99:5] error: expected a source argument for `mul` instruction (takes 2 arguments)

[../tests/ui/validation.wa:100:3-100:6] error: expected a destination argument for `mull` instruction (takes 3 arguments)

[../tests/ui/validation.wa:100:3-100:6] error: expected a destination argument for `mull` instruction (takes 3 arguments)

[../tests/ui/validation.wa:100:3-100:6] error: expected a source argument for `mull` instruction (takes 3 arguments)

[../tests/ui/validation.wa:101:3-101:6] error: expected a destination argument for `mulu` instruction (takes 2 arguments)

[../tests/ui/validation.wa:101:3-101:6] error: expected a source argument for `mulu` instruction (takes 2 arguments)

[../tests/ui/validation.wa:102:3-102:7] error: expected a destination argument for `mullu` instruction (takes 3 arguments)

[../tests/ui/validation.wa:102:3-102:7] error: expected a destination argument for `mullu` instruction (takes 3 arguments)

[../tests/ui/validation.wa:102:3-102:7] error: expected a source argument for `mullu` instruction (takes 3 arguments)

[../tests/ui/validation.wa:103:3-103:5] error: expected 2 arguments for `mul` instruction, found 3 arguments

[../tests/ui/validation.wa:104:3-104:6] error: expected 3 arguments for `mull` instruction, found 4 arguments

[../tests/ui/validation.wa:105:3-105:6] error: expected 2 arguments for `mulu` instruction, found 3 arguments

[../tests/ui/validation.wa:106:3-106:7] error: expected 3 arguments for `mullu` instruction, found 4 arguments

[../tests/ui/validation.wa:109:3-109:5] error: expected 0 arguments for `nop` instruction, found 1 arguments

[../tests/ui/validation.wa:110:3-110:5] error: expected 0 arguments for `nop` instruction, found 2 arguments

[../tests/ui/validation.wa:113:1-113:15] error: duplicate `.static` section
[../tests/ui/validation.wa:9:1-9:15] note: previously declared here

[../tests/ui/validation.wa:120:1-120:13] error: the `.code` section must occur before the `.static` section

[../tests/ui/validation.wa:120:1-120:13] error: duplicate `.code` section
[../tests/ui/validation.wa:52:1-52:13] note: previously declared here

error: aborting due to 54 previous errors

